File: dart/angular2/lib/components/button/button_component.dart
library angular2_dart_examples.button_component;

import 'dart:html';
import 'package:angular2/angular2.dart';
import 'package:angular2_dart_examples/components/button/utilities_service.dart';

@Component(
	selector: 'ui-button',
	templateUrl: 'button-template.html',
	styleUrls: const ['button-style.css'],
	encapsulation: ViewEncapsulation.Native,
	providers: const [Utilities],
	directives: const [NgClass])
class ButtonComponent implements OnInit {
	static const String PREFIX = 'button--';

	Map<String, bool> classes = {};

	HtmlElement element;

	@Input() String name = 'button';
	@Input('class') String custom;
	@Input() bool disabled;
	
	@Input() String version;
	@Input() String color;
	@Input() dynamic info;

	@Output() EventEmitter<dynamic> procedure = new EventEmitter();

	ButtonComponent(ElementRef reference) {
		this.element = reference.nativeElement;
		window.console.log(this.element.attributes);
		classes['button'] = true;
	}

	void ngOnInit() {
		if(custom != null) custom.split(' ').forEach(addClass);
		
		disabled = disabled == null ? false : true;
		
		Utilities.forEach(
			element.attributes,
			(dynamic value, [String key, Map<String, dynamic> object])
				=> addClass(value, true),
			['version', 'color']
		);
		
		// Bound properties
		print('version: ${version}, color: ${color}, info:');
		window.console.log(this.info);
		
		// Trigger procedure
		if(procedure != null) procedure.emit(null);
	}

	@HostListener('click', const ['event'])
	void onClick([Event event]) {
		procedure.emit(null);
	}

	// Method to avoid not generated setter bug
	Event event() {
		return new Event('click');
	}

	void addClass(String name, [bool prefix = false]) {
		name = prefix ? PREFIX + name : name;
		if (!classes.containsKey(name)) {
			classes[name] = true;
		}
	}
}

File: typescript/angular2/components/button/src/button.component.ts
import {Component, ViewEncapsulation, OnInit, Input, Output, EventEmitter, HostListener} from 'angular2/core'
import {NgClass} from 'angular2/common'
import {Utilities} from './utilities.service'

"use strict";

@Component({
	selector: 'ui-button',
	templateUrl: 'src/button-template.html',
	styleUrls: ['src/button-style.css'],
	encapsulation: ViewEncapsulation.Native,
	providers: [Utilities],
	directives: [NgClass]})
export class ButtonComponent implements OnInit {
	static PREFIX: string = 'button--';
	private classes: Object;
	
	@Input() name: string = 'button';
	@Input() class: string;
	@Input() disabled: boolean;
	
	@Input() version: string;
	@Input() color: string;
	@Input() info: any;
	
	@Output() procedure: EventEmitter<any> = new EventEmitter();
	
	constructor() {
		this.classes = {'button': true};
	}
	
	ngOnInit() {
		var self = this;
		
		if(this.class) this.class.split(' ').forEach((value) => this.addClass(value));
		
		this.disabled = this.disabled === undefined ? false : true;
		
		Utilities.forEach(this, function(value, key) {
			self.addClass(value, true);
		}, ['version', 'color']);
		
		// Bound properties
		console.log('version: ' + this.version + ', color: ' + this.color + ', info:', this.info);
		
		// Trigger procedure
		if(this.procedure) this.procedure.emit(null);
	}
	
	@HostListener('click', ['$event'])
	private onClick(event: Event): void {
		this.procedure.emit(null);
	}
	
	private addClass(name: string, prefix: boolean = false): void {
		name = prefix ? ButtonComponent.PREFIX + name : name;
		if (!this.classes[name]) {
			this.classes[name] = true;
		}
	}
}